### ISSUE TITLE
Instance tags in instance metadata availability

### ISSUE THREAD

**Issue by sergkondr** [aws/karpenter#3650]
### Tell us about your request

Hi Karpenter team!

As far as you do not allow configuring custom tags in the `provisioner` resource(#3025), we have to workaround it. And it would be very nice to have the ability to get tags [directly from instance metadata](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/)

If I get right, t [is not supported now](https://github.com/aws/karpenter/blob/0f26150297e706f8aad51b7572a0623f82b9e890/pkg/apis/v1alpha1/provider.go#L80)

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

We would like to use the scheme with one AWSNodeTemplate and separated provisioners, but we'd like to use different tags for instances created from different provisioners, because we think about provisioners as node groups.

### Are you currently working around this issue?

Yes, by api call `describe-tags`

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by bwagner5** [comment]
tags propagated via instance metadata are limited and not allowed to have a `/` which is problematic for a lot of tagging formats, especially used in K8s. For example, the `kubernetes.io/cluster` tag would result in the instance not launching. 

Are you able to duplicate the AWSNodeTemplates and include tags within there?

**Comment by sergkondr** [comment]
Hello! O, I didn't know that, thank you!

Yes, we are able to duplicate AWSNodeTemplates, but we'd like to avoid this, because, it looks like it is possible to solve our issue without it. And it looks not so convenient to maintain multiple identical resources with the only difference in one parameter.

BTW, do you have an issue to AWS about the format of the tags?

**Comment by dcarrion87** [comment]
This is a blocker for us and we're having to hack around as I can't seem to get Karpenter to enable this during instance launch.

**Comment by jonathan-innis** [comment]
> This is a blocker for us and we're having to hack around as I can't seem to get Karpenter to enable this during instance launch

Can you describe why this is a blocker for you? This is useful information that helps us communicate priority to the AWS team that supports this feature.

**Comment by RoseSecurity** [comment]
Just here to echo that this is a priority for us as well. We utilize a service that exposes the metadata tags as a label for other services to consume. The blocker is that Karpenter provisioned nodes do not disclose these tags, making it impossible for the additional services to access this information. We are architecting an auditing scheme that makes use of these tags, so this feature would be greatly appreciated.

**Comment by mikkoc** [comment]
We would love to have this, so we don't need to call describe-instance to retrieve the information about tags.

**Comment by OmryAr** [comment]
Adding our use case here as requested by maintainers.

**Use case: CrowdStrike Falcon Sensor host grouping**

We run CrowdStrike Falcon Sensor on all Karpenter-managed EKS nodes for endpoint security. Falcon Sensor supports automated host grouping based on EC2 instance metadata tags — this is the recommended way to dynamically assign sensor policies (e.g., different prevention/detection profiles for infra vs. app vs. compute nodes) without manual group management.

Without `instanceMetadataTags`, each node must fall back to `DescribeInstances` API calls to retrieve its own tags, which:
- Doesn't scale well across large fleets (we run ~900 EC2 instances across 6 regions in production)
- Requires granting broader IAM permissions to each node
- Adds latency to sensor initialization and policy assignment

**Additional blocked use cases on our side:**
- Compliance scanning tools running on nodes that need to self-identify their environment/role via IMDS tags
- Cost allocation agents that read `CostCenter`/`Team` tags directly from IMDS
- Instance self-registration scripts that use `Name`/`Environment` tags without needing IAM permissions for `ec2:DescribeInstances`

**Current workaround:** We have none for Karpenter nodes. Since PR #7012 force-disables `instanceMetadataTags`, there is no way to opt in — even for tags that are fully IMDS-compatible (no slashes). We've filed an AWS support case requesting that the EC2 IMDS endpoint gracefully skip tag keys with unsupported characters instead of rejecting the entire launch.

The ideal fix is on the AWS EC2 API side: allow `instanceMetadataTags=enabled` to coexist with slash-containing tags by simply omitting incompatible keys from the IMDS response rather than failing validation. This would unblock Karpenter from exposing the option and let customers access their own IMDS-compatible tags on Kubernetes nodes.
